src/BuiltInTools/BrowserRefresh/ScriptInjectingStream.cs

```diff
@@ -0,0 +1,378 @@
+        var minLength = Math.Min(buffer.Length, remainingBodyTagBytes.Length);
+
+        return buffer[..minLength].SequenceEqual(remainingBodyTagBytes[..minLength])
+            ? minLength
+            : -1;
+    }
+
+    private static int FindPartialTagLengthFromEnd(ReadOnlySpan<byte> buffer)
+    {
+        var bufferLength = buffer.Length;
+        if (bufferLength == 0)
+        {
+            return -1;
+        }
+
+        // Since each character within "</body>" is unique, we can use the last byte
+        // in the buffer to determine the length of the partial body tag.
+        var lastByte = buffer[^1];
+        var bodyMarkerIndexOfLastByte = BodyTagIndexOf(lastByte);
+        if (bodyMarkerIndexOfLastByte == -1)
+        {
+            // The last character does not appear "</body>", so we know
+            // there's not a partial body tag.
+            return -1;
+        }
+
+        var partialTagLength = bodyMarkerIndexOfLastByte + 1;
+        if (buffer.Length < partialTagLength)
+        {
+            // The buffer is shorter than the expected length of the partial
+            // body tag, so we know the buffer can't possibly contain it.
+            return -1;
+        }
+
+        // Finally, we need to check that the content at the end of the buffer
+        // matches the expected partial body tag.
+        return buffer[^partialTagLength..].SequenceEqual(s_bodyTagBytes.Span[..partialTagLength])
+            ? partialTagLength
+            : -1;
+
+        // We can utilize the fact that each character is unique in "</body>"
+        // to perform an efficient index lookup.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        static int BodyTagIndexOf(byte c)
+            => c switch
+            {
+                (byte)'<' => 0,
+                (byte)'/' => 1,
+                (byte)'b' => 2,
+                (byte)'o' => 3,
+                (byte)'d' => 4,
+                (byte)'y' => 5,
+                (byte)'>' => 6,
+                _ => -1,
+            };
+    }
+
+    public ValueTask CompleteAsync() => DisposeAsync();
+
+    protected override void Dispose(bool disposing)
+    {
+        if (disposing)
+        {
+            DisposeAsync().AsTask().GetAwaiter().GetResult();
+        }
+    }
+
+    public override async ValueTask DisposeAsync()
+    {
+        if (_isDisposed)
+        {
+            return;
+        }
+
+        _isDisposed = true;
+
+        if (_bodyTagBufferLength > 0)
+        {
+            // We might have buffered some data thinking that it could represent
+            // a body tag. We know at this point that there's no more data
+            // on its way, so we'll write the remaining data to the buffer.
+            await _baseStream.WriteAsync(_bodyTagBuffer.AsMemory()[.._bodyTagBufferLength]);
+            _bodyTagBufferLength = 0;
+        }
+
+        await FlushAsync();
+    }
```

Comment:
This problem of not being able to fully flush until disposal now appears in both `ScriptInjectingStream` and `ResponseStreamWrapper` :) But I think it's fine as it doesn't seem possible to make it behave otherwise, even in theory.